packages/docusaurus-theme-common/src/internal.ts

```diff
@@ -0,0 +1,129 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+// This re-export permits to handle some level of retro-compatibility
+// Users might swizzle unsafe components, using our internal apis.
+// When we move an API from internal to public, former internal imports
+// should keep working => less annoying for users.
+// TODO: how do we detect legacy internal imports in our own code?
+ 
+// export * from './index';
+
+export {
+  type Navbar,
+  type NavbarItem,
+  type NavbarLogo,
+} from './utils/useThemeConfig';
+export {
+  DocSidebarItemsExpandedStateProvider,
+  useDocSidebarItemsExpandedState,
+} from './contexts/docSidebarItemsExpandedState';
+export {DocsVersionProvider, useDocsVersion} from './contexts/docsVersion';
+export {DocsSidebarProvider, useDocsSidebar} from './contexts/docsSidebar';
+export {DocProvider, useDoc} from './contexts/doc';
+export type {DocContextValue} from './contexts/doc';
+
+export {createStorageSlot, listStorageKeys} from './utils/storageUtils';
+
+export {useAlternatePageUtils} from './utils/useAlternatePageUtils';
+
+export {
+  parseCodeBlockTitle,
+  parseLanguage,
+  parseLines,
+  containsLineNumbers,
+} from './utils/codeBlockUtils';
+
+export {docVersionSearchTag, DEFAULT_SEARCH_TAG} from './utils/searchUtils';
+
+export {
+  isDocsPluginEnabled,
+  useDocById,
+  findSidebarCategory,
+  findFirstCategoryLink,
+  isActiveSidebarItem,
+  useSidebarBreadcrumbs,
+  useDocsVersionCandidates,
+  useLayoutDoc,
+  useLayoutDocsSidebar,
+  useDocRouteMetadata,
+} from './utils/docsUtils';
+
+export {useTitleFormatter} from './utils/generalUtils';
+
+export {useLocationChange} from './utils/useLocationChange';
+
+export {useCollapsible, Collapsible} from './components/Collapsible';
+
+export {
+  useDocsPreferredVersion,
+  useDocsPreferredVersionByPluginId,
+  DocsPreferredVersionContextProvider,
+} from './contexts/docsPreferredVersion';
+
+export {duplicates, uniq} from './utils/jsUtils';
+
+export {
+  AnnouncementBarProvider,
+  useAnnouncementBar,
+} from './contexts/announcementBar';
+
+export {useLocalPathname} from './utils/useLocalPathname';
+
+export {useHistoryPopHandler} from './utils/historyUtils';
+
+export {
+  useTOCHighlight,
+  type TOCHighlightConfig,
+} from './hooks/useTOCHighlight';
+
+export {
+  useFilteredAndTreeifiedTOC,
+  useTreeifiedTOC,
+  type TOCTreeNode,
+} from './utils/tocUtils';
```

Comment:
At least `useTreeifiedTOC` is very useful

Reply:
`useTreeifiedTOC` is not used in any theme component, only `useFilteredAndTreeifiedTOC` is, which does feel like a weird name that we may want to change currently. Similarly, not sure i's useful to publicly commit to retro-compatibility on apis that even our own theme does not use. In the end we want the TOC components to be safe to swizzle, and in this case it doesn't help.

I would be ok to expose a unique `useTOCTree(array, options)` or something later, but this requires some refactoring.

I want to keep this PR clean from refactors: focusing only on splitting private/internal that we are ready to expose right now, as is.

What I mean is that we can do extra refactors in subsequent PRs, and it's not a big deal if this one API is not public immediately.

Reply:
I once planned to document `useTreeifiedTOC` in #6729, because it's no longer possible to pick arbitrary TOC subtrees so users need to implement this themselves. But I agree the API is not perfect yet and can benefit from some more refactors.